pallets/session-ext/src/manager.rs

```diff
@@ -0,0 +1,124 @@
+/// * New session is planned:
+/// 1. Inner T::new_session invoked
+/// 2. If session starts era EM::on_new_era invoked
+/// 3. Logic related to new session from this pallet is invoked
+/// * Session ends:
+/// 1. Inner T::end_session invoked
+/// 2. Logic related to new session from this pallet is invoked
+/// * Session starts:
+/// 1. Inner T::start_session invoked
+/// 2. Logic related to new session from this pallet is invoked
+/// 3. If session starts era EM::new_era_start invoked
+/// 4. If session starts era logic related to new era from this pallet is invoked
+pub struct SessionManagerExt<E, EM, T, C>(PhantomData<(E, EM, T, C)>)
+where
+    T: SessionManager<C::AccountId>,
+    EM: EraManager,
+    E: EraInfoProvider,
+    C: Config;
+
+impl<E, EM, T, C> SessionManagerExt<E, EM, T, C>
+where
+    T: SessionManager<C::AccountId>,
+    E: EraInfoProvider,
+    EM: EraManager,
+    C: Config,
+{
+    fn session_starts_era(session: SessionIndex, next_era: bool) -> Option<EraIndex> {
+        let mut active_era = match E::active_era() {
+            Some(ae) => ae,
+            // no active era, session can't start it
+            _ => return None,
+        };
+        if next_era {
+            active_era += 1;
+        }
+        if let Some(era_start_index) = E::era_start_session_index(active_era) {
+            return if era_start_index == session {
+                Some(active_era)
+            } else {
+                None
+            };
+        }
+
+        None
+    }
+}
+
+impl<E, EM, T, C> SessionManager<C::AccountId> for SessionManagerExt<E, EM, T, C>
+where
+    T: SessionManager<C::AccountId>,
+    E: EraInfoProvider,
+    EM: EraManager,
+    C: Config,
+{
+    fn new_session(new_index: SessionIndex) -> Option<Vec<C::AccountId>> {
+        T::new_session(new_index);
+        if let Some(era) = Self::session_starts_era(new_index, true) {
+            EM::on_new_era(era);
+            Pallet::<C>::emit_fresh_bans_event();
+        }
+
+        Pallet::<C>::rotate_committee(new_index)
+    }
+
+    fn end_session(end_index: SessionIndex) {
+        T::end_session(end_index);
+        Pallet::<C>::adjust_rewards_for_session();
+        Pallet::<C>::calculate_underperforming_validators();
+        // clear block count after calculating stats for underperforming validators, as they use
+        // SessionValidatorBlockCount for that
+        let result = SessionValidatorBlockCount::<C>::clear(u32::MAX, None);
+        debug!(target: "pallet_elections", "Result of clearing the `SessionValidatorBlockCount`, {:?}", result.deconstruct());
+    }
+
+    fn start_session(start_index: SessionIndex) {
+        T::start_session(start_index);
+        Pallet::<C>::clear_underperformance_session_counter(start_index);
+
+        if let Some(era) = Self::session_starts_era(start_index, false) {
+            EM::new_era_start(era);
+            Pallet::<C>::update_validator_total_rewards(era);
+            Pallet::<C>::clear_expired_bans(era);
```

Comment:
Shouldn't this be before `EM::new_era_start(era);`? In previous version, we checked expiring bans during elections.

Reply:
no longer relevant after changes